rep-xxxx.rst

```diff
@@ -0,0 +1,268 @@
+
+
+Specification
+=============
+
+Coordinate Frames
+-----------------
+
+base_link
+'''''''''
+
+The coordinate frame called ``base_link`` shall be rigidly attached to the robot root body.
+For industrial manipulators, this frame is expected to be located at the bottom of the first link - the base of the robot, or the surface that forms the robot's base mount - and forms the root of the kinematic chain of the manipulator.
+
+Note that this frame does not need to be coincident with the origin of the controller's Cartesian frame, but rather it should be located in the logical base position of the robot.
+
+This frame shall follow ROS conventions for orientation of frames as set forth in REP 103 [#REP103]_.
+Discrepancies between a controller's internal Cartesian frame and REP 103 [#REP103]_ shall be resolved by defining the correct transform between ``base_link`` (or some other convenient frame) and the ``base`` frame (see next section).
+
+``base_link`` may have geometry assigned to it, but this is not required.
+
+Rationale: this is the standard ROS ``base_link`` frame, but used in the context of industrial robots.
+No special treatment of this frame is required, but making it the root of the kinematic chain is strongly recommended.
+
+
+base
+''''
+
+The ``base`` frame shall be coincident with the default Cartesian coordinate system as defined by the industrial robot controller.
+Its purpose is to allow users to transform poses from a ROS application into the Cartesian base frame of the robot.
+As such, this frame is exempt from the requirement to follow orientation conventions as described in REP 103 [#REP103]_.
+
+Examples of vendor-specific names for this frame are *World* (Fanuc), *ROBROOT* (KUKA) and *Base* (ABB, Denso, Mitsubishi and Yaskawa Motoman).
+
+Any frame is acceptable as the parent of ``base`` (so not just ``base_link``), as long as the transform between parent and ``base`` is fixed (i.e.: not across a movable joint).
+
+``base`` shall not have any geometry associated with it.
+
+Finally: ``base`` may be part of the (main) kinematic chain, but this is not required.
+Frame hierarchies must however support transforms from anywhere in the chain to ``base`` and vice-versa (for example: from ``base`` to ``tool0``, or from ``link_4`` to ``base``).
+
+Rationale: for many industrial robots, the location of the ROS ``base_link`` frame is not necessarily coincident with the origin of their main Cartesian coordinate system.
+This complicates expressing poses in robot-relative frames, as arbitrary transformations could be required for different makes and models of robots.
+The ``base`` frame is intended to be used as an intermediary or bridge and provides a mapping that links the ROS TF tree with the controller- and vendor-specific coordinate systems and as an abstraction that decreases coupling between applications and specific robot urdfs.
+
+
+flange
+''''''
```

Comment:
Is the addition of the flang link just so there is link with the ros convention at the end of the kinematic change?

Reply:
Partly, yes.

But the rationale for having a link that is compliant with REP-103 is that it makes attaching EEF subhierarchies very convenient (ROS 'attaches' subtrees to x+ by default). `tool0` has z+ forward, which leads to your EEF model ending up rotated in weird / unexpected ways wrt your flange.

To fix that you'd have to undo the rotation that `tool0` introduces, which is what I've observed ppl do. By introducing `flange` we avoid all that.

Reply:
And it resolves the strange mixup of a physical attachment point (`flange`) and the mathematical concept that is the tool centre point (`tool0`) by separating them into two distinct frames.